thin-replica-server/src/replica_state_snapshot_service_impl.cpp

```diff
@@ -0,0 +1,105 @@
+// Concord
+//
+// Copyright (c) 2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "thin-replica-server/replica_state_snapshot_service_impl.hpp"
+
+#include "categorization/kv_blockchain.h"
+#include "Logger.hpp"
+#include "rocksdb/native_client.h"
+
+#include <stdexcept>
+#include <string>
+
+namespace concord::thin_replica {
+
+using vmware::concord::replicastatesnapshot::StreamSnapshotRequest;
+using vmware::concord::replicastatesnapshot::StreamSnapshotResponse;
+
+using bftEngine::impl::DbCheckpointManager;
+using kvbc::categorization::KeyValueBlockchain;
+using storage::rocksdb::NativeClient;
+
+grpc::Status ReplicaStateSnapshotServiceImpl::StreamSnapshot(grpc::ServerContext* context,
+                                                             const StreamSnapshotRequest* request,
+                                                             grpc::ServerWriter<StreamSnapshotResponse>* writer) {
+  const auto snapshot_id_str = std::to_string(request->snapshot_id());
+  if (!overriden_path_.has_value()) {
+    const auto checkpoint_state = overriden_checkpoint_state_.has_value()
+                                      ? *overriden_checkpoint_state_
+                                      : DbCheckpointManager::instance().getCheckpointState(request->snapshot_id());
+    switch (checkpoint_state) {
+      case DbCheckpointManager::CheckpointState::kNonExistent: {
+        const auto msg = "State Snapshot ID = " + snapshot_id_str + " doesn't exist";
+        LOG_INFO(STATE_SNAPSHOT, msg);
+        return grpc::Status{grpc::StatusCode::NOT_FOUND, msg};
+      }
+      case DbCheckpointManager::CheckpointState::kPending: {
+        const auto msg = "State Snapshot ID = " + snapshot_id_str + " is pending creation";
+        LOG_INFO(STATE_SNAPSHOT, msg);
+        return grpc::Status{grpc::StatusCode::UNAVAILABLE, msg};
+      }
+      case DbCheckpointManager::CheckpointState::kCreated:
```

Comment:
Let's cover the `default` case.

Reply:
I explicitly left it out so that I can take advantage of the compiler check for unhandled cases. If a new enum value is added, we will get an error that way.